c2rust-analyze/src/borrowck/type_check.rs

```diff
@@ -273,58 +303,31 @@ impl<'tcx> TypeChecker<'tcx, '_> {
-                    let adt_metadata = self.adt_metadata;
-
                     // Here we relabel `expect_ty` to utilize the permissions it carries
                     // but substitute the rest of its `Label`s' parts with fresh origins
                     // Othwerise, this is conceptually similar to labeling the cast target
                     // `ty`. We would simply do that, but do not have the information necessary
                     // to set its permissions.
-                    self.ltcx.relabel(expect_ty, &mut |lty| {
-                        let perm = lty.label.perm;
-                        match lty.ty.kind() {
-                            TyKind::Ref(_, _, _) | TyKind::RawPtr(_) => {
-                                let origin = Some(self.maps.origin());
-                                Label {
-                                    origin,
-                                    origin_params: &[],
-                                    perm,
-                                }
-                            }
-                            TyKind::Adt(..) => {
-                                let origin_params = construct_adt_origins(
-                                    &self.ltcx,
-                                    adt_metadata,
-                                    &lty.ty,
-                                    self.maps,
-                                );
-                                Label {
-                                    origin: None,
-                                    origin_params,
-                                    perm,
-                                }
-                            }
-                            _ => Label {
-                                origin: None,
-                                origin_params: &[],
-                                perm,
-                            },
-                        }
-                    })
+                    self.relabel_fresh_origins(expect_ty)
                 } else {
                     panic!("Creating non-null pointers from exposed addresses not supported");
                 }
             }
-            Rvalue::Cast(_, _, ty) => self.ltcx.label(ty, &mut |_ty| {
+            Rvalue::Cast(..) => {
                 // TODO: handle Unsize casts at minimum
                 /*
                 assert!(
                     !matches!(ty.kind(), TyKind::RawPtr(..) | TyKind::Ref(..)),
                     "pointer Cast NYI"
                 );
                 */
-                Label::default()
-            }),
+                // Here we relabel `expect_ty` to utilize the permissions it carries
+                // but substitute the rest of its `Label`s' parts with fresh origins
+                // Othwerise, this is conceptually similar to labeling the cast target
+                // `ty`. We would simply do that, but do not have the information necessary
+                // to set its permissions.
+                self.relabel_fresh_origins(expect_ty)
```

Comment:
I think we probably need to establish some kind of relationships between the fresh origins and the `Operand` of the `Rvalue::Cast`.  Otherwise, in `let z: Z = (x: X) as Y;`, there is a relationship between Y and Z (due to the enclosing `Assign`), but no relationship between X and Y, so X and Y/Z will be considered totally independent.

The null pointer case this is based on is a bit of a special case since it's effectively generating a pointer from thin air rather than deriving a new pointer from an existing one.  A better reference is the `Aggregate` case below, which (at least for tuples) generates some fresh origins but then connects them to existing operands.

Reply:
I think `do_assign` happens to mostly work, but it would be better to match on specific `CastKind`s/`TyKind`s and explicitly do the right thing for each one.  (`do_assign` should probably also assert that the two `Ty`s are equivalent, like `dataflow::type_check`'s `do_assign`/`do_equivalence_nested` does.)

Reply:
Is f3751839 sufficient for now? I think it covers the `lighttpd` failure and checks for equality. I think it might be best if we match on specific `CastKind`s/`TyKind`s as needed.

Reply:
Yes, that seems fine for now.  Though it might be slightly cleaner to do the `TyKind::RawPtr` check inside the `MutToConstPointer`, where it's clear why that logic is needed.